calicoctl/calicoctl/commands/ipam/split.go

```diff
@@ -0,0 +1,246 @@
+		oldPoolCIDR = parsedArgs["--cidr"].(string)
+	}
+	if name := parsedArgs["--name"]; name != nil {
+		oldPoolName = parsedArgs["--name"].(string)
+	}
+	if oldPoolCIDR == "" && oldPoolName == "" {
+		return fmt.Errorf("No name or CIDR provided. Provide a name or CIDR to denote the IP pool to split.")
+	}
+
+	cf := parsedArgs["--config"].(string)
+	// Get the backend client.
+	client, err := clientmgr.NewClient(cf)
+	if err != nil {
+		return err
+	}
+
+	// Check that the datastore is locked
+	ctx := context.Background()
+	locked, err := common.CheckLocked(ctx, client)
+	if err != nil {
+		return fmt.Errorf("Error while checking if datastore was locked: %s", err)
+	} else if !locked {
+		return fmt.Errorf("Datastore is not locked. Run the `calicoctl datastore migrate lock` command in order split the IP pools.")
+	}
+
+	// Find the IP pool to split
+	var oldPool *apiv3.IPPool
+	if oldPoolName != "" {
+		oldPool, err = client.IPPools().Get(ctx, oldPoolName, options.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("Unable to find IP pool with name %s: %v", oldPoolName, err)
+		}
+	} else if oldPoolCIDR != "" {
+		poolList, err := client.IPPools().List(ctx, options.ListOptions{})
+		if err != nil {
+			return fmt.Errorf("Unable to list IP pools to find the pool specified by %s", oldPoolCIDR)
+		}
+
+		for _, pool := range poolList.Items {
+			if pool.Spec.CIDR == oldPoolCIDR {
+				oldPool = &pool
+			}
+		}
+	}
+
+	if oldPool == nil {
+		return fmt.Errorf("Unable to find IP pool %s covering the specified CIDR %s", oldPoolName, oldPoolCIDR)
+	}
+
+	// Disable the specified IP pool.
+	oldPool.Spec.Disabled = true
+	oldPool, err = client.IPPools().Update(ctx, oldPool, options.SetOptions{})
+	if err != nil {
+		return fmt.Errorf("Error disabling IP pool %s", oldPoolCIDR)
+	}
+
+	// Calculate the split pool CIDRs.
+	numString := parsedArgs["<NUMBER>"].(string)
+	splitNum, err := strconv.Atoi(numString)
+	if err != nil {
+		return fmt.Errorf("Error reading number to split IP pools into. %s is not a valid number: %v", numString, err)
+	}
+
+	splitCIDRs, err := splitCIDR(oldPool.Spec.CIDR, splitNum)
+	if err != nil {
+		return fmt.Errorf("Error splitting the CIDR %s into %d CIDRs: %v", oldPool.Spec.CIDR, splitNum, err)
+	}
+
+	// Create the new split pools using UnsafeCreate.
+	poolsCreated := make([]*apiv3.IPPool, splitNum)
+	for i, cidr := range splitCIDRs {
+		poolCopy := oldPool.DeepCopyObject()
+		splitPool, ok := poolCopy.(*apiv3.IPPool)
+		if !ok {
+			return fmt.Errorf("Error copying metadata out from old IP pool: %s", oldPool.GetObjectMeta().GetName())
+		}
+
+		// Clear out unneeded metadata
+		splitPool.GetObjectMeta().SetUID("")
+		splitPool.GetObjectMeta().SetResourceVersion("")
+		splitPool.GetObjectMeta().SetGeneration(0)
+		splitPool.GetObjectMeta().SetSelfLink("")
+		splitPool.GetObjectMeta().SetCreationTimestamp(metav1.Time{})
+
+		// Set the new name
+		newName := fmt.Sprintf("split-%s-%d", oldPool.GetObjectMeta().GetName(), i)
```

Comment:
We might need to do something smarter here - k8s objects have a name character limit, and adding a prefix / suffix to an existing name might put us above that limit.

We need logic that says "if the name is > limit, then remove some characters and add a uniqueness marker to it".

This can be done in a follow-on PR though so long as you track it.